oci/archive/oci_src.go

```diff
@@ -0,0 +1,153 @@
+package archive
+
+import (
+	"context"
+	"io"
+
+	"os"
+
+	ocilayout "github.com/containers/image/oci/layout"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/pkg/archive"
+	digest "github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type ociArchiveImageSource struct {
+	ref         ociArchiveReference
+	unpackedSrc types.ImageSource
+	tempDir     string
+}
+
+var tempDirRef *tempDirOciRef
```

Comment:
This can’t be a global variable; multiple `ociArchiveImageSource` instances can exist at the same time. The necessary data can probably live inside the `ociArchiveImage$something`, or maybe just as local variables.

Reply:
>  However, two instances were being created and ociArchiveReference wouldn't retain its value.

Yeah,`ociArchiveReference` is defined to be a _value_ of that `struct` type, i.e. every value of type `ociArchiveReference` is an independent copy of the value, not a reference to a shared underlying object. I.e. every method with a receiver of `(ref ociArchiveReference)` gets a fresh copy, and every parameter with type `ociArchiveReference` (notably the `ref` parameters to `newImage{Source,Destination}`) gets a copy. You are probably seeing the copies made during the `newImage{Source,Destination}` parameter passing.

In general, `ImageReference` is mostly being treated as an immutable-after-construction object, so storing state in there is a bit tricky; though at least caching information in it for speed up (without changing semantics) should be fine.  The storage backend uses references that way.

OTOH attaching the lifetime of the temporary directories to `ImageReference` is, with the current structure, problematic, because multiple `ImageSource`/`ImageDestination` objects could be created from the same `ImageReference` over time (or even concurrently!), and there is no good place to delete the temporary directory: `Image{Source,Destination}.Close` would delete it after the first user is done, possibly breaking all other users, and there is no `ImageReference.Close`.

So, yeah, extracting the archive in `LoadManifestDescriptor`, and tearing down the extracted archive immediately in that function, and then extracting it again in `NewImageSource`, is the best fit with the interface structure. Arguably that’s just evidence of a badly designed interface which can’t deal with this reasonable use case… but the cost of making the `LoadManifestDescriptor+NewImageSource` sequence performant (forcing every user of `ImageReference` to treat them as stateful objects and to call `ref.Close` eventually, and dealing with multiple users and locking and reference counts in the `ImageReference` implementations) is fairly high.

Overall having `LoadManifest` extract the archive one more time seems like a reasonable compromise. Or, perhaps better, it could just seek inside the archive to get the manifest file, like `tarfile.Source.readTarComponent` does, without extracting gigabytes of data onto the disk; with an uncompressed archive this would be quite cheap. With a compressed archive we might still have to decompress gigabytes of data, but at least we wouldn’t have to write it to disk.

Perhaps we could also make a special-case exception, by documenting that a reference passed to `LoadManifestDescriptor` must be _always_, and _exactly once_, passed to `ref.NewImageSource()`, and implementing the pair of these calls as a special case, without dealing with the general case of multiple, possibly concurrent, callers to `ref.NewImage{Source,Destination}`. That would be quite ugly, but ugly only locally, without impacting the rest of the code, and high-performing. I guess I’d start with the extracting-twice solution (or streaming-and-extracting solution) first, and defer implementing this special case only if the performance improvement turned out to be necessary.